### ISSUE TITLE
"addend too big for relocation"

### ISSUE THREAD

**Issue by ilg-ul** [iains/gcc-darwin-arm64#52]
While compiling my packages on Apple Silicon, I occasionally encountered errors like this:

```console
gcc-xbb -DHAVE_CONFIG_H -I. -I/Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/sources/pixman-0.40.0/test -I..  -I/Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/sources/pixman-0.40.0/pixman -I../pixman -I/Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/install/libs/include/libpng16 -I/Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/install/libs/include -I/Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/install/libs/include  -fopenmp  -ffunction-sections -fdata-sections -pipe -O2 -w -c -o affine-bench.o /Users/ilg/Work/qemu-arm-2.8.0-13/darwin-arm64/sources/pixman-0.40.0/test/affine-bench.c
<stdin>:593:2: error: addend too big for relocation
        adrp    x1, _clean_space.0@PAGE+8388608
        ^
<stdin>:595:2: error: addend too big for relocation
        add     x1, x1, _clean_space.0@PAGEOFF+8388608;momd
        ^
<stdin>:595:2: error: fixup value out of range
        add     x1, x1, _clean_space.0@PAGEOFF+8388608;momd
        ^
make[2]: *** [Makefile:1161: affine-bench.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

In some cases simply updating to the latest version of the package made the error go away, but in this case even with the latest release of pixman, and using the same `configure` flags as on HomeBrew, the issue persists.

The problem was also reported on stackoverflow, about some HomeBrew builds:

- https://stackoverflow.com/questions/69696931/addend-too-big-for-relocation-error-reported-by-homebrew-g-installed-on-mac

Is this something related to GCC 11.1? If necessary, I can recreate my build environment with 11.2, if it is stable now.

**Comment by iains** [comment]
> While compiling my packages on Apple Silicon, I occasionally encountered errors like this:

> * https://stackoverflow.com/questions/69696931/addend-too-big-for-relocation-error-reported-by-homebrew-g-installed-on-mac

@fxcoudert 
Why has this not been reported here?

I need a proper reproducer case to fix the bug <gripe> complaining about bugs won't make them go away, they need fixing .. </gripe>

> Is this something related to GCC 11.1? If necessary, now I can recreate my build environment with 11.2, if it is stable now.

does your branch have this patch?:

https://github.com/iains/gcc-darwin-arm64/commit/2a516dc89274662e5ff8ddd4d868871487b9f1ef

(there is actually an assembler/link bug which is not representing negative values properly).

If the value is actually positive (which I can't tell without the reproducer) then that patch needs to be extended (I had a version that range-checked at one stage - but it never fired on any of the code in GCC).

**Comment by ilg-ul** [comment]
> does your branch have this patch?:

If the question is for me, I don't know, I'm using the HomeBrew branch.

**Comment by fxcoudert** [comment]
> Why has this not been reported here?

How should I know? I don't think it was reported to Homebrew either. I wasn't aware of it.

> does your branch have this patch?: 2a516dc

Yes it does.

> I need a proper reproducer case

I can reproduce with my patched 11.2:

```
meau /tmp $ cat a.cpp 
#include <bits/stdc++.h>
using namespace std;
vector<int> vec[1145140];
int main(){
    return 0;
}
meau /tmp $ g++-11 a.cpp 
/var/folders/h8/9hx_fyj91053ksgdzb2w03vw0000gp/T//ccJZqeGt.s:431:2: error: addend too big for relocation
        adrp    x0, _vec@PAGE+27483360
        ^
/var/folders/h8/9hx_fyj91053ksgdzb2w03vw0000gp/T//ccJZqeGt.s:432:2: error: addend too big for relocation
        add     x19, x0, _vec@PAGEOFF+27483360;momd
        ^
/var/folders/h8/9hx_fyj91053ksgdzb2w03vw0000gp/T//ccJZqeGt.s:432:2: error: fixup value out of range
        add     x19, x0, _vec@PAGEOFF+27483360;momd
        ^
```

The assembler produced can be found here: https://gist.github.com/fxcoudert/09f649d2613ae39a8c0f025d13f3a2df

**Comment by iains** [comment]
sigh, I've had a week of finding out that bugs have been reported everywhere but where I can see them ... sorry for being 'grumpy'.  Let's take a look at that (it might be a code model problem too - given the array size).

**Comment by fxcoudert** [comment]
@iains I can feel your pain, I'm trying to help out users on Homebrew forums as well. We had a report of failure to allocate big chunks of static memory here https://github.com/Homebrew/discussions/discussions/2520#discussioncomment-1704321 but it seems to be a generic limitation, since I can reproduce that in C with clang.

**Comment by iains** [comment]
this bug is confirmed (issue 52), I have an idea how to fix it .. but not tried yet.

----

I suspect that discussions#2520 relates to code model - presumably since the static allocation is 2Gb that exceeds the addressability of small code model.  IIRC x86_64 had this problem too (still does on GCC, but I think clang is fixed).

-- you might want to investigate if there's a large code model for aarch64-darwin (I did not implement one so far).

**Comment by ilg-ul** [comment]
> I have an idea how to fix it .

Great!

In the mean time I also had an idea how to avoid the bug... I compiled `pixman` with Apple clang instead of GCC  ;-)

**Comment by iains** [comment]
fixed by 040bb280be93df51c344c0720c241b658609ac61
